pyqir-generator/src/python.rs

```diff
@@ -1,174 +1,264 @@
-    Single,
+
+use crate::{
+    emit,
+    interop::{
+        ClassicalRegister, Controlled, If, Instruction, Measured, QuantumRegister, Rotated,
+        SemanticModel, Single,
+    },
 };
-use log;
-use pyo3::exceptions::PyOSError;
-use pyo3::prelude::*;
+use pyo3::{exceptions::PyOSError, prelude::*};
 
 #[pymodule]
-fn pyqir_generator(_py: Python<'_>, m: &PyModule) -> PyResult<()> {
-    m.add_class::<PyQIR>()?;
-    Ok(())
+fn _native(_py: Python, m: &PyModule) -> PyResult<()> {
+    m.add_class::<SimpleModule>()?;
+    m.add_class::<Qubit>()?;
+    m.add_class::<Ref>()?;
+    m.add_class::<Builder>()?;
+    m.add_class::<BasicQisBuilder>()
 }
 
+const RESULT_NAME: &str = "result";
+const QUBIT_NAME: &str = "qubit";
+
 #[pyclass]
-pub struct PyQIR {
-    pub(super) model: SemanticModel,
+struct SimpleModule {
+    model: SemanticModel,
+    builder: Py<Builder>,
 }
 
 #[pymethods]
-impl PyQIR {
+impl SimpleModule {
     #[new]
-    fn new(name: String) -> Self {
-        PyQIR {
-            model: SemanticModel::new(name),
+    fn new(py: Python, name: String, num_qubits: u64, num_results: u64) -> PyResult<SimpleModule> {
+        let registers = vec![ClassicalRegister::new(RESULT_NAME.to_string(), num_results)];
+
+        let qubits = (0..num_qubits)
+            .map(|i| QuantumRegister::new(QUBIT_NAME.to_string(), i))
+            .collect();
+
+        let model = SemanticModel {
+            name,
+            registers,
+            qubits,
+            instructions: Vec::new(),
+            static_alloc: true,
+        };
+
+        let builder = Py::new(py, Builder::new())?;
+        Ok(SimpleModule { model, builder })
+    }
+
+    #[getter]
+    fn qubits(&self) -> Vec<Qubit> {
+        self.model
+            .qubits
+            .iter()
+            .map(|q| Qubit { index: q.index })
+            .collect()
+    }
+
+    #[getter]
+    fn results(&self) -> Vec<Ref> {
+        let size = self.model.registers.first().unwrap().size;
+        (0..size)
+            .map(|index| Ref(RefKind::Result { index }))
+            .collect()
+    }
+
+    #[getter]
+    fn builder(&self) -> Py<Builder> {
+        self.builder.clone()
+    }
+
+    fn ir(&mut self, py: Python) -> PyResult<String> {
+        self.update_instructions(py);
+        emit::ir(&self.model).map_err(PyOSError::new_err)
+    }
+
+    fn bitcode(&mut self, py: Python) -> PyResult<Vec<u8>> {
+        self.update_instructions(py);
+        emit::bitcode(&self.model).map_err(PyOSError::new_err)
+    }
+}
+
+impl SimpleModule {
+    fn update_instructions(&mut self, py: Python) {
+        let builder = self.builder.as_ref(py).borrow();
+
+        match builder.frames[..] {
+            [ref instructions] => self.model.instructions = instructions.clone(),
+            _ => panic!("Builder does not contain exactly one stack frame."),
         }
     }
+}
+
+#[pyclass]
+struct Qubit {
+    index: u64,
+}
+
+impl Qubit {
+    fn id(&self) -> String {
```

Comment:
Should this be exposed as a `__repr__` method?

Reply:
This particular string should not be exposed - it's an implementation detail that may be removed in the future (the `SemanticModel` doesn't really need to use string identifiers). However, I implemented a different repr for Qubit and Ref:

```
<Qubit 0>
<Ref to Result 0>
```